Kuadrant/authorino
# Reconciliation errors reported in the status

Makes reconciliation errors and intermediate states to be reflected in the status sub-resource of the AuthConfigs.

Instead of relying on the state of the cache to make inferences about readiness, the status updater now shares another exclusive data structure with the reconciler for reporting errors, intermediate and final states.

This has 2 side effects:
1. The summary added to the status sub-resource is no longer a mix of previous state and current state. Now the number of hosts linked to the AuthConfig (available) should be based on the latest status report filled by the reconciler after the last attempt to reconcile the resource; therefore, phantom hosts (see #326) and hosts still cached while the reconciliation to the latest generation of the resource are not counted in the status summary. 
2. Replicas of the Authorino pod that are not elected the leader will fill status reports despite never consumed by the status updater (deactivated for reconciliation events in non-leader replicas).

Updated table of reasons for an `AuthConfig` related to the status condition `Ready`:

| Reason           | Description | Ready? | New? |
|------------------|-------------|:------:|:----:|
| `Reconciling`    | Transient state between reconciliation started and finished (due to success or terminal failure) | N | Y |
| `Invalid`        | Invalid resource. AuthConfig cannot be added to cache for any of the hosts listed. Message will reflect the exact error (e.g. broken ref). | N | Y |
| `HostsNotLinked` | One or more hosts not linked to the resource | N | N |
| `CachingError`   | Failed to add the AuthConfig to the cache | N | Y |
| `Reconciled`     | Resource reconciled | Y | N |
| `Unknown`        | Should never happen :-) | N | Y |

### Verification steps

Build, deploy, setup:

```sh
make local-setup # clusterWide: true

kubectl create namespace user-1
kubectl create namespace user-2

kubectl -n user-1 apply -f -<<EOF
apiVersion: authorino.kuadrant.io/v1beta1
kind: AuthConfig
metadata:
  name: auth-1
  namespace: user-1
spec:
  hosts:
  - talker-api.127.0.0.1.nip.io
EOF
```

Check the status for the one valid AuthConfig created:

```sh
kubectl get authconfigs -A -o wide
# NAMESPACE   NAME     READY   HOSTS   AUTHENTICATION   METADATA   AUTHORIZATION   RESPONSE   WRISTBAND
# user-1      auth-1   true    1/1     0                0          0               0          false
```

```sh
kubectl -n user-1 get authconfig/auth-1 -o jsonpath={.status}
# {"conditions":[{"lastTransitionTime":"2022-07-20T11:20:55Z","reason":"HostsLinked","status":"True","type":"Available"},{"lastTransitionTime":"2022-07-20T11:20:55Z","reason":"Reconciled","status":"True","type":         "Ready"}],"summary":{"festivalWristbandEnabled":false,"hostsReady":["talker-api.127.0.0.1.nip.io"],"numAuthorizationPolicies":0,"numHostsReady":"1/1","numIdentitySources":0,"numMetadataSources":0,"numResponseItems":0,   "ready":true}}
```

AuthConfig with host already taken by another:

```sh
kubectl -n user-2 apply -f -<<EOF
apiVersion: authorino.kuadrant.io/v1beta1
kind: AuthConfig
metadata:
  name: auth-2
  namespace: user-2
spec:
  hosts:
  - talker-api.io
  - talker-api.127.0.0.1.nip.io
EOF
```

```sh
kubectl get authconfigs -A -o wide
# NAMESPACE   NAME     READY   HOSTS   AUTHENTICATION   METADATA   AUTHORIZATION   RESPONSE   WRISTBAND
# user-1      auth-1   true    1/1     0                0          0               0          false
# user-2      auth-2   false   1/2     0                0          0               0          false

```

```sh
kubectl -n user-2 get authconfig/auth-2 -o jsonpath={.status}
# {"conditions":[{"lastTransitionTime":"2022-07-20T11:29:28Z","reason":"HostsLinked","status":"True","type":"Available"},{"lastTransitionTime":"2022-07-20T14:30:18Z","message":"One or more hosts not linked to the resource","reason":"HostsNotLinked","status":"False","type":"Ready"}],"summary":{"festivalWristbandEnabled":false,"hostsReady":["talker-api.io"],"numAuthorizationPolicies":0,"numHostsReady":"1/2","numIdentitySources":0,"numMetadataSources":0,"numResponseItems":0,"ready":false}}
```

Invalid AuthConfig:

```sh
kubectl -n user-2 apply -f -<<EOF
apiVersion: authorino.kuadrant.io/v1beta1
kind: AuthConfig
metadata:
  name: auth-2
  namespace: user-2
spec:
  hosts:
  - talker-api.io
  - talker-api.127.0.0.1.nip.io
  metadata:
  - name: userinfo
    userInfo:
      identitySource: foo # broken ref
EOF
```

```sh
kubectl get authconfigs -A -o wide
# NAMESPACE   NAME     READY   HOSTS   AUTHENTICATION   METADATA   AUTHORIZATION   RESPONSE   WRISTBAND
# user-1      auth-1   true    1/1     0                0          0               0          false
# user-2      auth-2   false   0/2     0                1          0               0          false
```

```sh
kubectl -n user-2 get authconfig/auth-2 -o jsonpath={.status}
# {"conditions":[{"lastTransitionTime":"2022-07-20T14:33:08Z","message":"No hosts linked to the resource","reason":"HostsNotLinked","status":"False","type":"Available"},{"lastTransitionTime":"2022-07-20T14:30:18Z","message":"Missing identity config foo","reason":"Invalid","status":"False","type":"Ready"}],"summary":{"festivalWristbandEnabled":false,"hostsReady":[],"numAuthorizationPolicies":0,"numHostsReady":"0/2","numIdentitySources":0,"numMetadataSources":1,"numResponseItems":0,"ready":false}}
```

Perfectly valid second AuthConfig:

```sh
kubectl -n user-2 apply -f -<<EOF
apiVersion: authorino.kuadrant.io/v1beta1
kind: AuthConfig
metadata:
  name: auth-2
  namespace: user-2
spec:
  hosts:
  - talker-api.io
EOF
```

```sh
kubectl get authconfigs -A -o wide
# NAMESPACE   NAME     READY   HOSTS   AUTHENTICATION   METADATA   AUTHORIZATION   RESPONSE   WRISTBAND
# user-1      auth-1   true    1/1     0                0          0               0          false
# user-2      auth-2   true    1/1     0                0          0               0          false
```

```sh
kubectl -n user-2 get authconfig/auth-2 -o jsonpath={.status}
# {"conditions":[{"lastTransitionTime":"2022-07-20T14:34:42Z","reason":"HostsLinked","status":"True","type":"Available"},{"lastTransitionTime":"2022-07-20T14:34:42Z","reason":"Reconciled","status":"True","type":"Ready"}],"summary":{"festivalWristbandEnabled":false,"hostsReady":["talker-api.io"],"numAuthorizationPolicies":0,"numHostsReady":"1/1","numIdentitySources":0,"numMetadataSources":0,"numResponseItems":0,"ready":true}}
```
